src/libopensc/card-cac.c

```diff
@@ -0,0 +1,1580 @@
+	cac_object_list_entry_t *tail;
+} cac_object_list_t;
+
+/*
+ * Flags for Current Selected Object Type
+ *   CAC files are TLV files, with TL and V separated. For generic
+ *   containers we reintegrate the TL anv V portions into a single 
+ *   file to read. Certs are also TLV files, but pkcs15 wants the
+ *   actual certificate. At select time we know the patch which tells
+ *   us what time of files we want to read. We remember that type
+ *   so that read_binary can do the appropriate processing.
+ */
+#define CAC_OBJECT_TYPE_CERT		1
+#define CAC_OBJECT_TYPE_TLV_FILE	4
+
+/*
+ * CAC private data per card state
+ */
+typedef struct cac_private_data {
+	int object_type;		/* select set this so we know how to read the file */
+	int cert_next;			/* index number for the next certificate found in the list */
+	u8 *cache_buf;			/* cached version of the currently selected file */
+	size_t cache_buf_len;		/* length of the cached selected file */
+	int cached;			/* is the cached selected file valid */
+	cac_cuid_t cuid;                /* card unique ID from the CCC */
+	u8 *cac_id;                     /* card serial number */
+	size_t cac_id_len;              /* card serial number len */
+	cac_object_list_t pki_list;              /* list of pki containers */
+	cac_object_list_entry_t *pki_current;    /* current pki object  _ctl function */
+	cac_object_list_t general_list;          /* list of general containers */
+	cac_object_list_entry_t *general_current;/* current object for _ctl function */
+} cac_private_data_t;
+
+#define CAC_DATA(card) ((cac_private_data_t*)card->drv_data)
+
+static cac_private_data_t *cac_new_private_data(void)
+{
+	cac_private_data_t *priv;
+	priv = calloc(1, sizeof(cac_private_data_t));
+	/* set other fields as appropriate */
+
+	return priv;
+}
+
+static void cac_free_object_list(struct cac_object_list *list);
+
+static void cac_free_private_data(cac_private_data_t *priv)
+{
+	cac_free_object_list(&priv->pki_list);
+	cac_free_object_list(&priv->general_list);
+	free(priv);
+	return;
+}
+
+/*
+ * Object list operations
+ */
+static int cac_add_object_to_list(cac_object_list_t *list, const cac_object_t *object)
+{
+	cac_object_list_entry_t *entry = malloc(sizeof(cac_object_list_entry_t));
+
+	if (entry == NULL) {
+		return SC_ERROR_OUT_OF_MEMORY;
+	}
+	entry->next = NULL;
+	entry->obj = *object;
+	if (list->head == NULL) {
+		list->head = list->tail = entry;
+		return SC_SUCCESS;
+	}
+	list->tail->next = entry;
+	list->tail = entry;
+	return SC_SUCCESS;
+}
+
+static void cac_free_object_list(struct cac_object_list *list)
+{
+	cac_object_list_entry_t *current, *next;
+
+	for (current = list->head; current; current=next) {
+		next = current->next;
+		free(current);
+	}
+}
+
+
+/*
+ * Set up the normal CAC paths
+ */
+#define CAC_STRING_AND_LENGTH(x)  x, sizeof(x)-1
+#define CAC_TO_AID(x) CAC_STRING_AND_LENGTH(x)
+#define CAC_TO_PATH_VALUE(x) CAC_STRING_AND_LENGTH(x)
+#define CAC_NULL_PATH { 0 }, 0
```

Comment:
We can simply get rid of `CAC_TO_PATH_VALUE` which is used only once and `CAC_NULL_PATH`, which is just static replacement (committed already).

But I am still not completely convinced about the `CAC_TO_AID` and how to make it "better". I can rewrite

    { CAC_TO_AID(CAC_1_RID "\x01\x00") }

to

    #define CAC_AID_LEN 7
    { CAC_1_RID "\x01\x00", CAC_AID_LEN }

with hard-coded length would be more error-prone in case we would do some modification in these paths. Frank, would this work for you or could we agree on leaving this one in the code in current state?